File: 프로그래머스/3/42892. 길 찾기 게임/길 찾기 게임.java
import java.util.Arrays;
import java.util.ArrayList;
import java.util.List;

class Solution {
    public int[][] solution(int[][] nodeinfo) {
        int[][] answer = new int[2][nodeinfo.length];
        /**
        *   이진트리가 반드시 구성되는 인풋이 주어진다.
        *   따라서 우선순위큐로 레벨이 낮은 원소를 기준으로 정렬하고, 이 순서대로 순회하며 트리를 구성한다.
        *       1. y값 기준 내림차순
        *       2. x값 기준 오름차순
        */
        
        Node[] nodes = new Node[nodeinfo.length];
        for (int i = 0; i < nodeinfo.length; i++) {
            Node node = new Node(i + 1, nodeinfo[i][0], nodeinfo[i][1]);
            nodes[i] = node;
        }
        Arrays.sort(nodes, (n1, n2) -> {
            if (n1.y == n2.y) {
                return n1.x - n2.x;
            }
            return n2.y - n1.y;
        });
        
        // 최우선순위 노드는 반드시 루트 노드이므로
        Node root = nodes[0];
        for (int i = 1; i < nodes.length; i++) {
            dfs(root, nodes[i]);            
        }
        
        List<Integer> pre = new ArrayList<>();
        List<Integer> post = new ArrayList<>();
        preorder(root, pre);
        postorder(root, post);
        
        answer[0] = pre.stream().mapToInt(Integer::intValue).toArray();
        answer[1] = post.stream().mapToInt(Integer::intValue).toArray();
        
        return answer;
    }
    
    void dfs(Node root, Node node) {
        if (root.x < node.x) {
            if (root.right == null) {
                root.right = node;
            } else {
                dfs(root.right, node);            
            }
        } else {
            if (root.left == null) {
                root.left = node;
            } else {
                dfs(root.left, node);
            }
        }
    }
    
    void preorder(Node node, List<Integer> pre) {
        pre.add(node.value);
        if (node.left != null) {
            preorder(node.left, pre);
        }
        if (node.right != null) {
            preorder(node.right, pre);
        }
    }
    
    void postorder(Node node, List<Integer> post) {
        if (node.left != null) {
            postorder(node.left, post);
        }
        if (node.right != null) {
            postorder(node.right, post);
        }
        post.add(node.value);
    }
    
    class Node {
        Node left, right;
        int value, x, y;
    
        public Node(int value, int x, int y) {
            this.value = value;
            this.x = x;
            this.y = y;
        }
        
        @Override
        public String toString() {
            return this.value + ": (" + this.x + ", " + this.y + ")";
        }
    }
}

File: 프로그래머스/3/42892. 길 찾기 게임/길 찾기 게임.cpp
#include <bits/stdc++.h>

using namespace std;

struct Node {
    int x, y, value;
    
    Node* left;
    Node* right;
    
    Node(int x, int y, int value): x(x), y(y), value(value), left(nullptr), right(nullptr) {}
};

struct Tree {
    Node* root;
    
    Tree() : root(nullptr) {}
    
    void insert(Node* node) {
        insertRecursive(root, node);
    }
    
    void insertRecursive(Node*& currNode, Node* node) {
        if (currNode == nullptr) {
            currNode = node;
            return;
        }
        
        if (currNode->x < node->x) {
            insertRecursive(currNode->right, node);
        } else {
            insertRecursive(currNode->left, node);
        }
    }
    
    void preorder(Node* node, vector<int>& result) {
        if (node == nullptr) {
            return;
        }
        
        result.push_back(node->value);
        preorder(node->left, result);
        preorder(node->right, result);
    }
    
    void postorder(Node* node, vector<int>& result) {
        if (node == nullptr) {
            return;
        }
        
        postorder(node->left, result);
        postorder(node->right, result);
        result.push_back(node->value);
    }
};

vector<vector<int>> solution(vector<vector<int>> nodeinfo) {
    vector<vector<int>> answer;
    /**
    *   트리를 구성하는 노드들은 다음과 같은 순서로 부모 - 자식 순서 및 좌우 위치가 결정된다.
    *   1. y값이 큰 노드가 작은 노드보다 부모로 위치한다.
    *   2. x값이 부모 노드보다 작은 노드는 부모 노드의 왼쪽 자식 노드로 위치한다.
    *   위 기준을 기준으로 각 노드의 적절한 위치를 찾아 매번 말단 노드를 추가한다. -> O(N logN)
    *   트리의 전위/후위 순회를 기록한다. -> O(N)
    */
    
    vector<Node> nodes;
    for (int i = 0; i < nodeinfo.size(); i++) {
        nodes.push_back(Node(nodeinfo[i][0], nodeinfo[i][1], i + 1));
    }
    
    sort(nodes.begin(), nodes.end(), [](const auto& a, const auto& b) {
        if (a.y == b.y) {
            // x값 기준 오름차순
            return a.x < b.x;
        }
        // y값 기준 내림차순
        return a.y > b.y;
    });
    Tree tree;
    
    for (auto& node: nodes) {
        tree.insert(&node);
    }
    
    vector<int> preorderResult;
    vector<int> postorderResult;
    
    tree.preorder(tree.root, preorderResult);
    tree.postorder(tree.root, postorderResult);
    
    answer.push_back(preorderResult);
    answer.push_back(postorderResult);
    
    return answer;
}